t/socket-tcp-setoption.t

```diff
@@ -0,0 +1,725 @@
+        content_by_lua_block {
+            require "resty.core.socket"
+
+            local port = ngx.var.port
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect("127.0.0.1", port)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+
+            ok, err = sock:setoption()
+            if not ok then
+                ngx.say("setoption failed: ", err)
+            end
+
+            ok, err = sock:setoption("sndbuf")
+            if not ok then
+                ngx.say("setoption failed: ", err)
+            end
+
+            sock:close()
+        }
+    }
+--- request
+GET /t
+--- response_body
+setoption failed: missing the "option" argument
+setoption failed: missing the "value" argument
+--- no_error_log
+[error]
+
+
+
+=== TEST 2: unsuppotrted option name.
+--- config
+    set $port $TEST_NGINX_SERVER_PORT;
+
+    location /t {
+        content_by_lua_block {
+            require "resty.core.socket"
+
+            local port = ngx.var.port
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect("127.0.0.1", port)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+
+            local ok, err = sock:setoption("abc", 123)
+            if not ok then
+                ngx.say("setoption abc failed: ", err)
+                return
+            end
+
+            sock:close()
+        }
+    }
+--- request
+GET /t
+--- response_body
+setoption abc failed: unsupported option abc
+--- no_error_log
+[error]
+
+
+
+=== TEST 3: getoption before calling connect.
+--- config
+    set $port $TEST_NGINX_SERVER_PORT;
+
+    location /t {
+        content_by_lua_block {
+            require "resty.core.socket"
+
+            local sock = ngx.socket.tcp()
+            local sndbuf, err = sock:setoption("so_sndbuf", 4000)
+            if not sndbuf then
+                ngx.say("getoption so_sndbuf failed: ", err)
+                return
+            end
+
+            sock:close()
+        }
+    }
+--- request
+GET /t
+--- error_code: 500
+--- error_log
+bad tcp socket
+
+
+
+=== TEST 4: keepalive set by 1/0.
+--- config
+    set $port $TEST_NGINX_SERVER_PORT;
+
+    location /t {
+        content_by_lua_block {
+            require "resty.core.socket"
+
+            local port = ngx.var.port
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect("127.0.0.1", port)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+            local v1, err = sock:getoption("keepalive")
+            if not v1 then
+                ngx.say("get default keepalive failed: ", err)
+                return
+            end
+
+            ok, err = sock:setoption("keepalive", 1)
+            if not ok then
+                ngx.say("enable keepalive failed: ", err)
```

Comment:
Grammar: "enable" => "enabling". Fix all the other similar places.

Reply:
accept